src/server/pkg/testutil/env.go

```diff
@@ -1,48 +1,52 @@
 package testutil
 
 import (
+	"context"
 	"fmt"
 	"io/ioutil"
+	"net"
+	"net/url"
 	"os"
 	"path"
 
 	etcd "github.com/coreos/etcd/clientv3"
 	"github.com/coreos/etcd/embed"
+	"github.com/coreos/pkg/capnslog"
+	"golang.org/x/sync/errgroup"
+
 	"github.com/pachyderm/pachyderm/src/client"
+	authserver "github.com/pachyderm/pachyderm/src/server/auth/server"
+	authtesting "github.com/pachyderm/pachyderm/src/server/auth/testing"
+	pfsserver "github.com/pachyderm/pachyderm/src/server/pfs/server"
+	"github.com/pachyderm/pachyderm/src/server/pkg/hashtree"
+	"github.com/pachyderm/pachyderm/src/server/pkg/serviceenv"
+	txnenv "github.com/pachyderm/pachyderm/src/server/pkg/transactionenv"
+	txnserver "github.com/pachyderm/pachyderm/src/server/transaction/server"
 )
 
-// Env contains the basic setup for running end-to-end pachyderm tests entirely
+// EtcdEnv contains the basic setup for running end-to-end pachyderm tests entirely
 // locally within the test process. It provides a temporary directory for
-// storing data, an embedded etcd server with a connected client, as well as a
-// local mock pachd instance which allows a test to hook into any pachd calls.
-type Env struct {
+// storing data, and an embedded etcd server with a connected client.
+type EtcdEnv struct {
```

Comment:
Broke up the test env into three layers:

`EtcdEnv`, the lowest layer, creates a local scratch directory, runs an embedded etcd instance in-process, and provides an etcd client connected to this instance.

`MockEnv` depends on `EtcdEnv` and adds a pachd server with all RPC endpoints mockable (but undefined by default, so they error).

`RealEnv` (open to suggestions on the name, here) depends on `MockEnv` and instantiates real instance of pachd API servers, then uses reflection to forward each of the mock RPCs in the `MockEnv` pachd to call the real API server instances.

Each of these envs is scoped in a callback-based function allowing us to ensure proper and orderly cleanup.  It also lets tests pick which level of instantiation they need, as well as wrapping one of these env types into a test-suite local version that performs other setup operations.